Capabilities_Report_Design.md

```diff
@@ -0,0 +1,112 @@
+
+### Goals
+
+  * Design a unified API interface to collect capabilities.
+  * Design a workflow of capabilities discovery.
+
+### Non-Goals
+
+  * Integration to different storage drivers.
+  * Gap analysis of capabilities report feature among different backends.
+
+## Proposal
+
+Currently [Hotpot](https://github.com/opensds/opensds) project contains two services: ```osdslet``` service works as a controller to provide unified storage management, in the meantime ```osdsdock``` service runs on all storage nodes as a distributed agent. Since the capabilities report mechanism is a service from bottom to up, the service process should run on storage nodes and be triggered in polling mode. The concrete design would be introduced below.
+
+### Data model impact
+
+The proposal could imporve some properties in ```PoolSpec```, but considering the current format of ```ExtraSpec``` is map[string]interface{}, there would be no impact to the data model.
+
+Here is a sample JSON object of updated ```PoolSpec```:
+```json
+{
+  "id": "f4486139-78d5-462d-a7b9-fdaf6c797e1b",
+  "createdAt": "2018-01-15T12:19:45.505Z",
+  "updatedAt": "2018-01-15T12:19:45.505Z",
+  "name": "pool01",
+  "storageType": "block",
+  "description": "",
+  "status": "available",
+  "availabilityZone": "az01",
+  "totalCapacity": 1000,
+  "freeCapacity": 699,
+  "dockId": "ccac4f33-e603-425a-8813-371bbe10566e",
+  "extras": {
+    "DataStorageLineOfService": {
+      "RecoveryTimeObjective":0, // both thin and thick are supported in a pool
+      "ProvisioningPolicy":["Thin", “Fixed”],
+      "IsSpaceEfficient":true
+	},
+	"IOConnectivityLineOfService": {
+      "AccessProtocol":"FC",
+      "MaxIOPs":null
+    },
+    "DataProtectionLineOfService": {},
+    "xxx": {}
+  }
+}
+```
+
+### REST API impact
+
+None
+
+### Security impact
+
+This proposal brings IPC mechanism, so we should take care of message synchronization and protect the data from loss if system crashes.
+
+***Q1***: If the main thread change the status of backend to ```error```, how can report thread accept this signal and jump out of the loop? What if the backend service comes alive again?
+***Q2***: If the main thread crashes down, how to clean the report thread and notify ```osdslet``` daemon? 
+
+### Other end user impact
+
+None
+
+### Performance impact
+
+None
+
+### Other deployer impact
+
+None
+
+### Developer impact
+
+None
```

Comment:
"Developer impact" means what a driver developer needs to do with regard to this change.  So in the case, driver developer will need to implement capabilities reporting in the driver.